system/jlib/jsocket.cpp

```diff
@@ -445,22 +449,120 @@ extern jlib_decl bool queryKeepAlive(int &time, int &intvl, int &probes)
                     keepAliveTime = keepalive->getPropInt("@time", keepAliveTime);
                     keepAliveInterval = keepalive->getPropInt("@interval", keepAliveInterval);
                     keepAliveProbes = keepalive->getPropInt("@probes", keepAliveProbes);
-                    state = ENABLED;
+                    hasKeepAlive = true;
                 }
+                disableDNSTimeout = expert->getPropBool("@disableDNSTimeout", false);
+                maxDNSThreads = expert->getPropInt("@maxDNSThreads", maxDNSThreads);
+                // could also consider maxDNSThreads==0 as a way to disable DNS timeout ...
+                if (maxDNSThreads < 1)
+                    maxDNSThreads = 1;
             }
-            doKeepAlive = state;
+            expertTCPSettings = INITED;
         }
     }
+}
 
-    if (state == ENABLED)
+extern jlib_decl bool queryKeepAlive(int &time, int &intvl, int &probes)
+{
+    queryTCPSettings();
+    if (hasKeepAlive)
     {
         time = keepAliveTime;
         intvl = keepAliveInterval;
         probes = keepAliveProbes;
+    }
+    return hasKeepAlive;
+}
+
+static bool getAddressInfo(const char *name, unsigned *netaddr, bool okToLogErr);
+
+static CriticalSection queryDNSCS;
+
+class CAddrInfoThreadArgs : public CInterface, implements IInterface
+{
+public:
+    IMPLEMENT_IINTERFACE;
+    StringAttr name;
+    unsigned netaddr[4] = { 0, 0, 0, 0 };
+    bool result = false;
+
+    CAddrInfoThreadArgs(const char *_name) : name(_name) { }
+};
+
+class CAddrPoolThread : public CInterface, implements IPooledThread
+{
+public:
+    IMPLEMENT_IINTERFACE;
+
+    Linked<CAddrInfoThreadArgs> localAddrThreadInfo;
+
+    virtual void init(void *param) override
+    {
+        localAddrThreadInfo.setown((CAddrInfoThreadArgs *)param);
+    }
+
+    virtual void threadmain() override
+    {
+        localAddrThreadInfo->result = getAddressInfo(localAddrThreadInfo->name.get(), localAddrThreadInfo->netaddr, false);
+    }
+
+    virtual bool stop() override
+    {
         return true;
     }
-    else
-        return false;
+
+    virtual bool canReuse() const override
+    {
+        return true;
+    }
+};
+
+class CAddrInfoFactory : public CInterface, public IThreadFactory
+{
+public:
+    IMPLEMENT_IINTERFACE;
+    IPooledThread *createNew()
+    {
+        return new CAddrPoolThread();
+    }
+};
+
+static Owned<CAddrInfoFactory> addrInfoFactory;
+static Owned<IThreadPool> addrInfoPool;
+
+static bool useDNSTimeout()
+{
+    queryTCPSettings();
+    if (!disableDNSTimeout)
+    {
+        if (!addrInfoPool)
+        {
+            CriticalBlock block(queryDNSCS);
+            if (!addrInfoPool)
```

Comment:
theoretically not thread safe since this isn't an atomic variable.  Probably the cleanest fix is to use a separate boolean atomic.

Reply:
added atomic,